brain/detail/circuit.h

```diff
@@ -222,12 +189,335 @@ class Circuit::Impl
+
+        if (subnetworks.size() > 1)
+        {
+            LBWARN << "More than one subnetwork found, ignored." << std::endl;
+        }
+
+        const auto populations = nodes.getPopulationNames();
+        if (populations.size() > 1)
+        {
+            LBWARN << "More than one population found, ignored." << std::endl;
+        }
+
+        const auto population = populations.front();
+        const auto nodeGroupIDs = nodes.getNodeGroupIDs(population);
+        const auto nodeIDs = nodes.getNodeIDs(population);
+        const auto nodeGroupIndices = nodes.getNodeGroupIndices(population);
+        const size_t numNodes = nodeIDs.size();
+
+        if (nodeGroupIDs.size() > 1)
+        {
+            LBWARN << "More than one group ID found, ignored." << std::endl;
+        }
+
+        nodeGroup = nodes.openGroup(population, 0);
+
+        size_t expectedIdx = 0;
+        for (size_t i = 0; i < numNodes; i++)
+        {
+            const auto groupID = nodeGroupIDs[i];
+            const auto nodeGroupIndex = nodeGroupIndices[i];
+
+            if (groupID != 0)
+                continue;
+
+            if (expectedIdx != nodeGroupIndex)
+            {
+                LBTHROW(
+                    std::runtime_error("Expected node group index '" +
+                                       std::to_string(expectedIdx) + "' got '" +
+                                       std::to_string(nodeGroupIndex) + "'"));
+            }
+
+            expectedIdx++;
+        }
+
+        numNeurons = expectedIdx;
+        morphologySource = URI(config.getComponentPath("morphologies_dir"));
+    }
+    virtual ~SonataCircuit() {}
+    virtual size_t getNumNeurons() const { return numNeurons; }
+    virtual GIDSet getGIDs(const std::string& /*target*/) const
+    {
+        LBUNIMPLEMENTED;
+        return GIDSet();
+    }
+
+    virtual Vector3fs getPositions(const GIDSet& gids) const
+    {
+        if (gids.empty())
+            return Vector3fs();
+
+        const size_t startIdx = *gids.begin();
+        const size_t endIdx = *gids.rbegin() + 1;
+
+        const auto attributeX =
+            nodeGroup.getAttribute<float>("x", startIdx, endIdx);
+        const auto attributeY =
+            nodeGroup.getAttribute<float>("y", startIdx, endIdx);
+        const auto attributeZ =
+            nodeGroup.getAttribute<float>("z", startIdx, endIdx);
+
+        Vector3fs output;
+        for (auto gid : gids)
+        {
+            const float x = attributeX[gid - startIdx];
+            const float y = attributeY[gid - startIdx];
+            const float z = attributeZ[gid - startIdx];
+
+            output.push_back(Vector3f(x, y, z));
+        }
+        return output;
+    }
+    virtual size_ts getMTypes(const GIDSet& /*gids*/) const
+    {
+        LBUNIMPLEMENTED;
+        return size_ts();
+    }
+    virtual Strings getMorphologyNames() const
+    {
+        return nodeGroup.getAttribute<std::string>("morphology_file");
```

Comment:
UNIMPLEMENTED, the name is a bit misleading, it's not the morphology file names, it's the morphology type names. Which we don't have in ABI's circuits as far as I know.